### PR TITLE
Parallel tool calling

### PR DESCRIPTION
**Pull request by aantn**
Improves latency by 9%, also reduces the maximum latency on outliers.

### CODE REVIEW

File: holmes/core/tool_calling_llm.py [hunk 1]
```diff
@@ -144,30 +146,40 @@ def call(self, system_prompt, user_prompt, post_process_prompt: Optional[str] =
...
+
+    def _invoke_tool(self, tool_to_call):
+        tool_name = tool_to_call.function.name
+        tool_params = json.loads(tool_to_call.function.arguments)
+        tool_call_id = tool_to_call.id
+        tool = self.tool_executor.get_tool_by_name(tool_name)
+        tool_response = tool.invoke(tool_params)
+        MAX_CHARS = 100_000     # an arbitrary limit - we will do something smarter in the future
+        if len(tool_response) > MAX_CHARS:
+            logging.warning(f"tool {tool_name} returned a very long response ({len(tool_response)} chars) - truncating to last 10000 chars")
```
Comment (arikalon1) [inline holmes/core/tool_calling_llm.py]: missing 0 in 10000

File: holmes/core/tool_calling_llm.py [hunk 2]
```diff
@@ -144,30 +146,40 @@ def call(self, system_prompt, user_prompt, post_process_prompt: Optional[str] =
...
+                        {
+                            "tool_call_id": tool_call_result.tool_call_id,
+                            "role": "tool",
+                            "name": tool_call_result.tool_name,
+                            "content": tool_call_result.result,
+                        }
                     )
-                )
+
+    def _invoke_tool(self, tool_to_call):
```
Comment (arikalon1) [inline holmes/core/tool_calling_llm.py]: Can you add the return type?
Also, what is the type of `tool_to_call`? What fields are in there?

**Review by arikalon1 — APPROVED** [review]
nice work

Left 2 minor comments

File: holmes/core/tool_calling_llm.py (hunk 2 above)
Comment (aantn) [reply holmes/core/tool_calling_llm.py]: Fixed

File: holmes/core/tool_calling_llm.py (hunk 1 above)
Comment (aantn) [reply holmes/core/tool_calling_llm.py]: Fixed

**Pull request merged**